src/pybind/mgr/nfs/export.py

```diff
@@ -314,6 +314,17 @@ def _create_export_user(self, export: Export) -> None:
             rgwfsal.secret_access_key = j['keys'][0]['secret_key']
             log.debug("Successfully fetched user %s for RGW path %s", rgwfsal.user_id, export.path)
 
+    def _ensure_cephfs_export_user(self, export: Export) -> None:
+        assert isinstance(export.fsal, CephFSFSAL)
+        fsal = cast(CephFSFSAL, export.fsal)
+        assert fsal.fs_name
+        assert fsal.cmount_path
+        fsal.user_id = f"nfs.{export.cluster_id}.{export.fsal.fs_name}"
+        fsal.cephx_key = self._create_user_key(
+            export.cluster_id, fsal.user_id, fsal.cmount_path, fsal.fs_name
+        )
```

Comment:
I think this change in the naming structure is causing a test failure
```
2024-02-09T13:08:07.529 INFO:tasks.cephfs_test_runner:======================================================================
2024-02-09T13:08:07.529 INFO:tasks.cephfs_test_runner:FAIL: test_create_and_delete_export (tasks.cephfs.test_nfs.TestNFS)
2024-02-09T13:08:07.529 INFO:tasks.cephfs_test_runner:Test successful creation and deletion of the cephfs export.
2024-02-09T13:08:07.529 INFO:tasks.cephfs_test_runner:----------------------------------------------------------------------
2024-02-09T13:08:07.529 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2024-02-09T13:08:07.529 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_2b25e10cc573cee6dfe2f5900f62ba4e83f9ac06/qa/tasks/cephfs/test_nfs.py", line 448, in test_create_and_delete_export
2024-02-09T13:08:07.529 INFO:tasks.cephfs_test_runner:    self._create_default_export()
2024-02-09T13:08:07.529 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_2b25e10cc573cee6dfe2f5900f62ba4e83f9ac06/qa/tasks/cephfs/test_nfs.py", line 231, in _create_default_export
2024-02-09T13:08:07.529 INFO:tasks.cephfs_test_runner:    self._create_export(export_id='1', create_fs=True)
2024-02-09T13:08:07.529 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_2b25e10cc573cee6dfe2f5900f62ba4e83f9ac06/qa/tasks/cephfs/test_nfs.py", line 219, in _create_export
2024-02-09T13:08:07.529 INFO:tasks.cephfs_test_runner:    self._check_auth_ls(export_id, check_in=True)
2024-02-09T13:08:07.529 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_2b25e10cc573cee6dfe2f5900f62ba4e83f9ac06/qa/tasks/cephfs/test_nfs.py", line 130, in _check_auth_ls
2024-02-09T13:08:07.530 INFO:tasks.cephfs_test_runner:    self.assertIn(f'{client_id}.{export_id}', output)
2024-02-09T13:08:07.530 INFO:tasks.cephfs_test_runner:AssertionError: 'client.nfs.test.1' not found in 'mds.a.smithi179.omhiad\n\tkey: AQDdIsZlQFS3BhAAa9Zq6Mxs1XZHa10jwIjUew==\n\tcaps: [mds] allow\n\tcaps: [mon] profile mds\n\tcaps: [osd] allow rw tag cephfs *=*\nmds.a.smithi179.zzaanp\n\tkey: AQDJIsZl2tIqFhAAYWbJxC9qls0g8Qd53rH0kw==\n\tcaps: [mds] allow\n\tcaps: [mon] profile mds\n\tcaps: [osd] allow rw tag cephfs *=*\nmds.nfs-cephfs.smithi179.khleme\n\tkey: AQCrI8Zl8QNNJBAAZZwPFhqdWRSrikOpQ9gysg==\n\tcaps: [mds] allow\n\tcaps: [mon] profile mds\n\tcaps: [osd] allow rw tag cephfs *=*\nmds.nfs-cephfs.smithi179.uzhtks\n\tkey: AQCuI8ZllR3eHBAA0VJRJDxyy1sU6tiBWgLitw==\n\tcaps: [mds] allow\n\tcaps: [mon] profile mds\n\tcaps: [osd] allow rw tag cephfs *=*\nosd.0\n\tkey: AQBWIsZlf4w1MBAAJ1yyxRe0pGhVk8e3awIpGw==\n\tcaps: [mgr] allow profile osd\n\tcaps: [mon] allow profile osd\n\tcaps: [osd] allow *\nosd.1\n\tkey: AQBwIsZlr084CxAAIEkAoTVPxY+21BvFNkSHaw==\n\tcaps: [mgr] allow profile osd\n\tcaps: [mon] allow profile osd\n\tcaps: [osd] allow *\nosd.2\n\tkey: AQCKIsZl5jGuAhAAWOUQRCca+AqOJAvlYNL6jw==\n\tcaps: [mgr] allow profile osd\n\tcaps: [mon] allow profile osd\n\tcaps: [osd] allow *\nclient.0\n\tkey: AQCnIsZlY6ILDBAAXkuOToR/WbjXrvwtsfdyug==\n\tcaps: [mds] allow *\n\tcaps: [mgr] allow *\n\tcaps: [mon] allow *\n\tcaps: [osd] allow *\nclient.admin\n\tkey: AQDwIcZlAOt3EhAA6xc5jZh6DnaqxIKTKbaD8g==\n\tcaps: [mds] allow *\n\tcaps: [mgr] allow *\n\tcaps: [mon] allow *\n\tcaps: [osd] allow *\nclient.bootstrap-mds\n\tkey: AQD4IcZlnWWZJhAA1gZBv62MOILxtl0NujyqaQ==\n\tcaps: [mon] allow profile bootstrap-mds\nclient.bootstrap-mgr\n\tkey: AQD4IcZluG2ZJhAA6Bfwi8zXxZev9+26g1iSiw==\n\tcaps: [mon] allow profile bootstrap-mgr\nclient.bootstrap-osd\n\tkey: AQD4IcZltHSZJhAASlE8fs4zdpstBZvzS6g/lw==\n\tcaps: [mon] allow profile bootstrap-osd\nclient.bootstrap-rbd\n\tkey: AQD4IcZl4nuZJhAAvItM4qkKoYDonawnWxwqAA==\n\tcaps: [mon] allow profile bootstrap-rbd\nclient.bootstrap-rbd-mirror\n\tkey: AQD4IcZl/IKZJhAA7vtWvpndXPu96QpMgHafmA==\n\tcaps: [mon] allow profile bootstrap-rbd-mirror\nclient.bootstrap-rgw\n\tkey: AQD4IcZlXIuZJhAAeFumY4Gq/QKfnMSB9WZOow==\n\tcaps: [mon] allow profile bootstrap-rgw\nclient.nfs.test.0.0.smithi179.pokwoy\n\tkey: AQChI8ZlRA5YIBAA4jlB9BpkAz5LlpId8VwJUw==\n\tcaps: [mon] allow r\n\tcaps: [osd] allow rw pool=.nfs namespace=test\nclient.nfs.test.0.0.smithi179.pokwoy-rgw\n\tkey: AQChI8ZlmZBlJhAA/jVuNnhQdqlQzcQIETxO/w==\n\tcaps: [mon] allow r\n\tcaps: [osd] allow rwx tag rgw *=*\nclient.nfs.test.nfs-cephfs\n\tkey: AQC2I8ZlVtjCGRAAO9aBG88+wjm1YC8nGi3W8w==\n\tcaps: [mds] allow rw path=/\n\tcaps: [mon] allow r\n\tcaps: [osd] allow rw pool=.nfs namespace=test, allow rw tag cephfs data=nfs-cephfs\nclient.test\n\tkey: AQAmI8ZlsbrfBhAAUcmdKo3prqtJXbJ9BKFMyg==\n\tcaps: [mds] allow rw path=/\n\tcaps: [mon] allow r\n\tcaps: [osd] allow rw pool=.nfs namespace=test, allow rw tag cephfs data=user_test_fs\nmgr.x\n\tkey: AQDxIcZl7MpbChAA7p3Hju/uwUdRxjX6m41Z4w==\n\tcaps: [mds] allow *\n\tcaps: [mon] profile mgr\n\tcaps: [osd] allow *\n'
```
It's looking for `client.nfs.test.1` but the keyring that's actually there is `client.nfs.test.nfs-cephfs` after this name change. The check comes from https://github.com/ceph/ceph/blob/main/qa/tasks/cephfs/test_nfs.py#L121 so I'd guess you'd have to adjust that to expect the new naming structure, although I'm not too familiar with the working s of the test myself.

https://pulpito.ceph.com/adking-2024-02-09_04:25:39-orch:cephadm-wip-adk-testing-2024-02-08-1828-distro-default-smithi/7553533/

Reply:
@adk3798  I've fixed the API tests for this. Can you re-trigger teuthology?

Reply:
@avanthakkar Changes to pybind/nfs would required reviews/acks from cephfs devs.

Reply:
@avanthakkar the test has gotten farther, but still seem to fail elsewhere
```
2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_635ee6850c57c98cc6ba932cade45f9cc015aa32/qa/tasks/cephfs/test_nfs.py", line 449, in test_create_and_delete_export
2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:    self._test_get_export()
2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_635ee6850c57c98cc6ba932cade45f9cc015aa32/qa/tasks/cephfs/test_nfs.py", line 285, in _test_get_export
2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:    self.assertDictEqual(self.sample_export, nfs_output)
2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:AssertionError: {'export_id': 1, 'path': '/', 'cluster_id':[219 chars]: []} != {'access_type': 'RW', 'clients': [], 'clust[251 chars]CP']}
2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:  {'access_type': 'RW',
2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:   'clients': [],
2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:   'cluster_id': 'test',
2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:   'export_id': 1,
2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:-  'fsal': {'fs_name': 'nfs-cephfs', 'name': 'CEPH', 'user_id': 'nfs.test.1'},
2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:+  'fsal': {'cmount_path': '/',
2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:+           'fs_name': 'nfs-cephfs',
2024-02-27T06:26:42.790 INFO:tasks.cephfs_test_runner:+           'name': 'CEPH',
2024-02-27T06:26:42.790 INFO:tasks.cephfs_test_runner:+           'user_id': 'nfs.test.nfs-cephfs'},
```
the user_id and cmount_path fields is not present in the dict its comparing to, and there is a difference in the user_id field. 
https://pulpito.ceph.com/adking-2024-02-27_05:08:20-orch:cephadm-wip-adk-testing-2024-02-26-1600-distro-default-smithi/7574464/
Note in that test run it also complains about a difference in the protocols, but that was caused by another PR, not this one.

Reply:
> @avanthakkar the test has gotten farther, but still seem to fail elsewhere
> 
> ```
> 2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:Traceback (most recent call last):
> 2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_635ee6850c57c98cc6ba932cade45f9cc015aa32/qa/tasks/cephfs/test_nfs.py", line 449, in test_create_and_delete_export
> 2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:    self._test_get_export()
> 2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:  File "/home/teuthworker/src/git.ceph.com_ceph-c_635ee6850c57c98cc6ba932cade45f9cc015aa32/qa/tasks/cephfs/test_nfs.py", line 285, in _test_get_export
> 2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:    self.assertDictEqual(self.sample_export, nfs_output)
> 2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:AssertionError: {'export_id': 1, 'path': '/', 'cluster_id':[219 chars]: []} != {'access_type': 'RW', 'clients': [], 'clust[251 chars]CP']}
> 2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:  {'access_type': 'RW',
> 2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:   'clients': [],
> 2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:   'cluster_id': 'test',
> 2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:   'export_id': 1,
> 2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:-  'fsal': {'fs_name': 'nfs-cephfs', 'name': 'CEPH', 'user_id': 'nfs.test.1'},
> 2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:+  'fsal': {'cmount_path': '/',
> 2024-02-27T06:26:42.789 INFO:tasks.cephfs_test_runner:+           'fs_name': 'nfs-cephfs',
> 2024-02-27T06:26:42.790 INFO:tasks.cephfs_test_runner:+           'name': 'CEPH',
> 2024-02-27T06:26:42.790 INFO:tasks.cephfs_test_runner:+           'user_id': 'nfs.test.nfs-cephfs'},
> ```
> 
> the user_id and cmount_path fields is not present in the dict its comparing to, and there is a difference in the user_id field. https://pulpito.ceph.com/adking-2024-02-27_05:08:20-orch:cephadm-wip-adk-testing-2024-02-26-1600-distro-default-smithi/7574464/ Note in that test run it also complains about a difference in the protocols, but that was caused by another PR, not this one.

Need to add `cmount_path` and change `user_id` in `self.sample_export` present in `def setUp()`

Reply:
FWIW, that `_create_user_key` perhaps has a misleading name. If the key already exists, (it use `ceph aut hget-or-create`) it won't actually create a new one, it will just return the existing one, which we might need to do anyway.

Reply:
> FWIW, that `_create_user_key` perhaps has a misleading name. If the key already exists, (it use `ceph auth get-or-create`) it won't actually create a new one, it will just return the existing one, which we might need to do anyway.

Oh, didn't check that code since the name was sort of self-explanatory. I think that func should be renamed to something like `_ensure_user_key`. Thanks @adk3798!